tests/test_client_enhanced_3.py

```diff
@@ -0,0 +1,1071 @@
+    expect(lambda: assert_that(rp, is_none()))
+    expect(lambda: assert_that(mock_finish.call_count, equal_to(1)))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data'], has_key('rerun_of')))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data']['rerun_of'], equal_to(rerun_id)))
+
+
+@mock.patch('reportportal_client.aio.client.requests.Session')
+def test_rp_client_set_attributes(self, mock_session):
+    mock_finish = mock_session().post
+    mock_finish.return_value = mock.Mock(json=lambda: {}, status_code=200)
+
+    client = RPClient('http://endpoint', 'default_personal', api_key='test_api_key',
+                      project='default_personal')
+    client.client = mock_session()
+    client.launch_uuid = TEST_LAUNCH_ID
+    client.set_test_attributes([{"key": "smoke", "value": "smoke"}])
+    expect(lambda: assert_that(mock_finish.call_count, equal_to(1)))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data'], has_key('attributes')))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data']['attributes'], has_length(1)))
+
+
+@mock.patch('reportportal_client.aio.client.requests.Session')
+def test_rp_client_set_custom_attributes(self, mock_session):
+    mock_finish = mock_session().post
+    mock_finish.return_value = mock.Mock(json=lambda: {}, status_code=200)
+
+    client = RPClient('http://endpoint', 'default_personal', api_key='test_api_key',
+                      project='default_personal')
+    client.client = mock_session()
+    client.launch_uuid = TEST_LAUNCH_ID
+    custom_attributes = [{"key": "smoke", "value": "smoke"}, {"value": "custom_attribute"}]
+    client.set_test_attributes(custom_attributes)
+    expect(lambda: assert_that(mock_finish.call_count, equal_to(1)))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data'], has_key('attributes')))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data']['attributes'], has_length(2)))
+
+
+@mock.patch('reportportal_client.aio.client.requests.Session')
+def test_rp_client_set_empty_attributes(self, mock_session):
+    mock_finish = mock_session().post
+    mock_finish.return_value = mock.Mock(json=lambda: {}, status_code=200)
+
+    client = RPClient('http://endpoint', 'default_personal', api_key='test_api_key',
+                      project='default_personal')
+    client.client = mock_session()
+    client.launch_uuid = TEST_LAUNCH_ID
+    client.set_test_attributes([])
+    expect(lambda: assert_that(mock_finish.call_count, equal_to(1)))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data'], has_key('attributes')))
+    expect(lambda: assert_that(mock_finish.call_args[1]['data']['attributes'], has_length(0)))
+
+
+@mock.patch('reportportal_client.aio.client.requests.Session')
+def test_rp_client_set_attributes_without_values(self, mock_session):
+    mock_finish = mock_session().post
+    mock_finish.return_value = mock.Mock(json=lambda: {}, status_code=200)
```

Comment:
A lot of lines like these. Need to collapse them to a function

As for the lines that do testing, they look better but I can't tell you if they are correct without launching them first